### PR TITLE
[CSL-1703] Fix a memory leak in Pos.Delegation.Logic.Common

### PR DESCRIPTION
**Pull request by adinapoli-iohk**
This PR fixes a memory leak in `Pos.Delegation.Logic.Common`, as reviewed together with @arybczak . To replicate this, I have isolated the piece of code I thought was leaking (I was lucky that it was a perfect self-contained candidate) and whipped up a small single-module which I could test with a fast turnaround. The program was replicating - in a pathological setting - the behaviour in Cardano, where we were invalidating the delegation cache by calling each second `runDelegationStateAction $ invalidateProxyCaches`.

This is the situation from which I started:

screen shot 2017-09-26 at 10 38 04

As you can see this exhibit the classic "exponential growth", whereas we would be expecting constant memory allocation. This is because we have X (say 100) slots available into the LRU cache, so after an initial allocation period we should see memory stabilise to the `maxBound` of the cache, but this wasn't the case. After adding the bang, this is how memory looks like:

screen shot 2017-09-26 at 10 42 27

Much better! The `PINNED` you see there is because I'm storing `ByteString`s into the cache.

## The icing on top

This chart looks fine, but I suggest we delve a little bit more into the `lrucache` package, aka we fork & maintain it up until we don't replace the current caching mechanisms. Maybe it's overkill, but there are some low-hanging fruits:

* [fromList](https://github.com/chowells79/lrucache/blob/master/src/Data/Cache/LRU/Internal.hs#L73) is using `foldr`, instead of its strict counterpart. I think that being lazy here might not be the best choice as ultimately we need to consume the entire list to build the final cache anyway.

* The library is not using `Data.Map.Strict` but its lazy version for most part, it seems for compatibility reasons, but I think we can expect in 2017 to have the majority of projects using a decently-updated version of `containers`. In particular, I feel that for LRU should apply the same principle of `Data.Map.Strict` which is that it should be _strict on the spine, lazy on the values_. Current implementation isn't, for no good reasons.

In the sandboxed experiments I've told you about I have done _exactly_ this, and this is the end result, which I like better because it yields a much "fuller" memory allocation pattern (but I might be wrong, so please GHC experts do chime in 😉  -- one may argue that the first graph is better because it's allocated on demand in only that spike, whereas this second one seems to yield a bit more predictable allocation pattern. No clue which one is best!) :

screen shot 2017-09-26 at 10 49 09

Last but not least: In Cardano we have the following function in `Pos.Util.LRU`:

```haskell
-- | Filter LRU cache by given predicate.
filterLRU :: Ord k => (v -> Bool) -> LRU.LRU k v -> LRU.LRU k v
filterLRU predicate lru =
    LRU.fromList (LRU.maxSize lru) . filter (predicate . snd) . LRU.toList $ lru
```

I think we can do much better if we modify the `LRU` type to be a `Functor` and even a `Bifunctor`, so we wouldn't need to do this "unpacking/packing" every time we want to operate on a cache.

### CODE REVIEW

**Comment by volhovM** [comment]
Regarding the suggestion to maintain the fork of `lrucache` I'd say that i see it as useless overhead. Isn't it better to make PR with suggested changes to https://github.com/chowells79/lrucache directly? 

By the way, remember that it is not possible to make `Bifunctor` instance of `LRUCache` for the same reason you can't do it with `Data.Set` -- methods require keys to be `Ord`'ed which doesn't match with `Bifunctor` class methods definition.

**Comment by adinapoli-iohk** [comment]
> Isn't it better to make PR with suggested changes to https://github.com/chowells79/lrucache directly?

The package seems abandoned, but of course if the maintainer is able to pick our fixes up, that's 1000 times better than forking.

> By the way, remember that it is not possible to make Bifunctor instance of LRUCache for the same reason you can't do it with Data.Set -- methods require keys to be Ord'ed which doesn't match with Bifunctor class methods definition.

Ah drat you are dead right, I realise just now that indeed there is the nagging `Ord` instance which prevents us from hitting the data structure with category theory 😬  but regardless, this shouldn't stop us from having an internal API (internal as it coded inside `Data.Cache.LRU.Internal`) and later exposed to application writers which would provide things like `map`, `filter` and so on and so forth (basically in the same vein as `Data.Set` exposes the various map, filter, etc combinators, which are implemented more efficiently than converting to & from list every time).

**Comment by adinapoli-iohk** [comment]
Apparently merging master wasn't enough. Trying to run `pkgs/generate.sh` as well..

File: node/src/Pos/Delegation/Logic/Common.hs [hunk 1]
```diff
@@ -78,7 +78,13 @@ runDelegationStateAction action = do
...
         v0 <- readTVar var
         let (r,v1) = runState action v0
-        writeTVar var v1
+        -- The `$!` is necessary here, otherwise `runDelegationStateAction` will
+        -- leak memory due to the fact `writeTVar` could end up accumulating thunks (if its
+        -- content is never demanded). Note that using a strict State is not enough,
+        -- as that ensure that only the sequencing of actions is strict, but the state
+        -- `s` is still lazy, so is using `Control.Lens.%=`, which calls `modify` and
+        -- not `modify'` under the hood.
+        writeTVar var $! v1
```
[the remark this answers is not in the document]
Comment (adinapoli-iohk) [reply node/src/Pos/Delegation/Logic/Common.hs]: The weird thing is that I think I did try that yesterday evening and, surprisingly, that was still leaking. Will try to confirm with my sandbox example tomorrow morning.

**Comment by dcoutts** [comment]
Ok @fersel, I'm happy for this to be merged.

File: stack.yaml [hunk 2]
```diff
@@ -22,6 +22,10 @@ packages:
 - explorer
 
 - location:
+    git: https://github.com/serokell/log-warper.git
+    # master, but we should switch to stackage when 1.2.2 will be on it
```
Comment (gromakovsky) [inline stack.yaml]: @volhovM can you explain what are we waiting for? Previously it was enough for a package to be on hackage. What changed?

Comment (volhovM) [reply stack.yaml]: I'm not sure it ever was as simple as you say. It seems that i'm wrong after all and we don't use `nixpkgs`' hackage clone to get packages we need so putting log-warper from hackage will work. I will switch it now.

**Comment by domenkozar** [comment]
This was merged to 1.0 branch.

**Comment by volhovM** [comment]
@domenkozar right, we'll merge cardano-sl-1.0 to master instead.

**Pull request closed without merging**